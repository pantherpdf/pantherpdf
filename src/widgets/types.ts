/**
 * @project PantherPDF Report Editor
 * @copyright Ignac Banic 2021-2023
 * @license MIT
 */

import type { IconDefinition } from '@fortawesome/fontawesome-svg-core';
import type { FunctionComponent } from 'react';
import type { ApiEndpoints, Name, ReportCompiled, Report } from '../types';
import type { FormulaHelper } from '../data/compile';
import type { GeneralProps } from '../editor/types';

/** Building blocks for editor */
export interface Widget {
  /** Unique widget id (type) */
  id: string;

  /** Human friendly name */
  name: Name;

  icon: IconDefinition;

  /** Create new instance of this widget */
  newItem: (props: ItemNewProps) => Promise<Item>;

  /** Compile item */
  compile: (item: any, helper: CompileHelper) => Promise<ItemCompiled>;

  /** Render properties editor */
  RenderProperties?: FunctionComponent<ItemRenderEditorProps>;

  /** Render item inside editor */
  RenderEditor: FunctionComponent<ItemRenderEditorProps>;

  /** Render item for final report */
  RenderPreview: FunctionComponent<ItemRenderPreviewProps>;

  /**
   * Can user add this widget to report? Default true.
   * Useful for helper Widgets which user should not manually manipulate.
   */
  canAdd?: boolean;

  /** Can user drag this widget to report (inside editor)? Default true. */
  canDrag?: boolean;

  /** Can user select Item by clicking ot it? Default true. */
  canSelect?: boolean;
}

/**
 * Instance of a Widget.
 * Reports are made out of multiple items. It contains data like text,
 * font-size, image url, margins etc.
 */
export interface Item {
  [key: string]: unknown;

  /** Id of a Widget */
  type: string;

  /** Subitems */
  children: Item[];
}

export interface ItemCompiled {
  [key: string]: unknown;
  type: string;
}

export interface ItemNewProps {
  report: Report;
}

export interface CompileHelper {
  formulaHelper: FormulaHelper;
  evalFormula: (txt: string) => Promise<unknown>;
  compileChildren: (
    children: Item[],
    helper: CompileHelper,
  ) => Promise<ItemCompiled[]>;

  /** Hierarchical index for current item */
  wid: number[];

  report: Report;
  reportCompiled: ReportCompiled;
  api: ApiEndpoints;
  externalHelpers: { [key: string]: any };
  variables: { [key: string]: unknown };
}

export interface ItemRenderEditorProps extends GeneralProps {
  item: Item;
  setItem: (itm: Item) => void;
  wid: number[];
}

export interface ItemRenderPreviewHelper {
  renderItem: (
    item: ItemCompiled,
    helper: ItemRenderPreviewHelper,
  ) => React.ReactNode;
  renderChildren: (
    chs: ItemCompiled[],
    helper: ItemRenderPreviewHelper,
  ) => React.ReactNode[];
  externalHelpers: { [key: string]: any };
}

export interface ItemRenderPreviewProps extends ItemRenderPreviewHelper {
  item: ItemCompiled;
}
