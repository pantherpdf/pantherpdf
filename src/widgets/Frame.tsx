/**
 * Frame.tsx
 */


import React, { CSSProperties } from 'react'
import { TData, TDataCompiled } from '../types'
import type { Widget } from '../editor/types'
import { faBorderStyle } from '@fortawesome/free-solid-svg-icons'
import PropertyColor from './PropertyColor'
import PropertyBorder, { Border, genBorderCss } from './PropertyBorder'
import BoxName from './BoxName'
import Trans from '../translation'
import InputApplyOnEnter, { WidthOptions, WidthRegex } from './InputApplyOnEnter'
import PropertyFont, { PropertyFontExtractStyle, PropertyFontGenCss, TFont } from './PropertyFont'
import { LoadGoogleFontCss } from './GoogleFonts'
import useStateDelayed from '../useStateDelayed'


interface Properties {
	margin: [number, number, number, number],
	padding: [number, number, number, number],
	border: Border | [Border, Border, Border, Border],
	width: string,
	height: string,
	backgroundColor?: string,
	pageBreakAvoid?: boolean,
	font: TFont,
}
export type FrameData = TData & Properties
export type FrameCompiled = TDataCompiled & Properties & {children: TDataCompiled[]}


function genStyle(item: FrameData | FrameCompiled, final: boolean): CSSProperties {
	const css: CSSProperties = {
		margin: `${item.margin[0]}px ${item.margin[1]}px ${item.margin[2]}px ${item.margin[3]}px`,
		padding: `${item.padding[0]}px ${item.padding[1]}px ${item.padding[2]}px ${item.padding[3]}px`,
		boxSizing: 'border-box',
		width: 'auto',  // in preview, to override width:100%
	}
	
	if (Array.isArray(item.border)) {
		css.borderTop = genBorderCss(item.border[0])
		css.borderRight = genBorderCss(item.border[1])
		css.borderBottom = genBorderCss(item.border[2])
		css.borderLeft = genBorderCss(item.border[3])
	}
	else {
		css.border = genBorderCss(item.border)
	}
	if (item.backgroundColor) {
		css.backgroundColor = item.backgroundColor
	}

	if (final) {
		if (item.width.length > 0) {
			css.width = item.width
			css.flex = `0 0 ${item.width}`
			css.overflowX = 'hidden'
		}
		if (item.height.length > 0) {
			css.height = item.height
			css.overflowY = 'hidden'
		}
	}
	else {
		if (item.width.length > 0) {
			css.width = item.width
		}
		if (item.height.length > 0) {
			css.minHeight = item.height
		}
	}

	if (item.pageBreakAvoid) {
		css.pageBreakInside = 'avoid'
	}

	const cssFont = PropertyFontGenCss(item.font)
	Object.assign(css, cssFont)

	return css
}


type Property4SideRangeValue = [number, number, number, number]
interface Property4SideRangeProps {
	id: string,
	label: string,
	min: number,
	max: number,
	value: Property4SideRangeValue,
	onChange: (val: Property4SideRangeValue) => void,
}
function Property4SideRange(props: Property4SideRangeProps) {
	const [value, setValue] = useStateDelayed<Property4SideRangeValue>(props.value, props.onChange)

	function renderInput(idx: number) {
		const st: CSSProperties = {
			width: '50%',
			display: 'inline',
		}
		function setValueInput(val: string, delay: number) {
			const arr: Property4SideRangeValue = [...value]
			arr[idx] = parseInt(val)
			return setValue(arr, delay)
		}
		return <input
			type='range'
			id={`${props.id}-${idx}`}
			style={st}
			min={props.min}
			max={props.max}
			value={value[idx]}
			onChange={e => setValueInput(e.currentTarget.value, 300)}
			onMouseUp={e => setValueInput(e.currentTarget.value, 0)}
			className='form-range'
		/>
	}

	return <div>
		<div className='section-name'>
			{props.label}
			<small className='ms-2 text-muted'>
				{value[0]}, {value[1]}, {value[2]}, {value[3]} px
			</small>
		</div>
		<div style={{textAlign:'center'}}>
			{renderInput(0)}
		</div>
		<div className="d-flex">
			{renderInput(3)}
			{renderInput(1)}
		</div>
		<div style={{textAlign:'center'}}>
			{renderInput(2)}
		</div>
	</div>
}




export const Frame: Widget = {
	name: {en: 'Frame', sl: 'Okvir'},
	icon: {fontawesome: faBorderStyle},

	newItem: async (): Promise<FrameData> => {
		return {
			type: 'Frame',
			children: [],
			margin: [0, 0, 0, 0],
			padding: [0, 0, 0, 0],
			border: {
				width: 1,
				style: 'solid',
				color: '#333333',
			},
			width: '',
			height: '',
			font: {},
		}
	},

	compile: async (dt: FrameData, helpers): Promise<FrameCompiled> => {
		const dt2: FrameCompiled = JSON.parse(JSON.stringify({...dt, children: []}))
		const style = PropertyFontExtractStyle(dt.font)
		if (style) {
			helpers.reportCompiled.fontsUsed.push(style)
		}
		dt2.children = await helpers.compileChildren(dt.children, helpers)
		return dt2
	},

	Render: function(props) {
		const item = props.item as FrameData
		const fontStyle = PropertyFontExtractStyle(item.font)
		if (fontStyle) {
			LoadGoogleFontCss(fontStyle)
		}
		return <BoxName {...props} style={genStyle(item, false)} name={Frame.name}>
			{props.renderWidgets(item.children, props.wid)}
		</BoxName>
	},

	RenderFinal: function(props) {
		const item = props.item as FrameCompiled
		const style = genStyle(item, true)
		return `<div style="${props.styleToStringAttribute(style)}">
			${props.renderChildren(item.children, props)}
		</div>`
	},

	RenderProperties: function(props) {
		const item = props.item as FrameData
		return <>
			<div className='vform'>
				<label htmlFor='Frame-width'>
					{Trans('width')}
					<small className='text-muted ms-1'>
						[{WidthOptions}]
					</small>
				</label>
				<InputApplyOnEnter
					id='Frame-width'
					value={item.width || ''}
					onChange={val => props.setItem({...item, width: String(val)})}
					regex={WidthRegex}
				/>
			</div>

			<div className='vform'>
				<label htmlFor='Frame-height'>
					{Trans('height')}
					<small className='text-muted ms-1'>
						[{WidthOptions}]
					</small>
				</label>
				<InputApplyOnEnter
					id='Frame-height'
					value={item.height || ''}
					onChange={val => props.setItem({...item, height: String(val)})}
					regex={WidthRegex}
				/>
			</div>

			<div className="form-check">
				<input
					type='checkbox'
					id='Frame-backgroundColor-enable'
					className='form-check-input'
					checked={!!item.backgroundColor}
					onChange={e => {
						const obj: FrameData = {...item}
						if (e.currentTarget.checked) {
							obj.backgroundColor = '#FFCCCC'
						}
						else {
							delete obj.backgroundColor
						}
						props.setItem(obj)
					}}
				/>
				<label className="form-check-label" htmlFor='Frame-backgroundColor-enable'>
					{Trans('background')}
				</label>
			</div>
			{!!item.backgroundColor && <PropertyColor
				value={item.backgroundColor}
				onChange={val => props.setItem({...item, backgroundColor: val})}
			/>}

			<Property4SideRange
				id='Frame-margin'
				label={Trans('margin')}
				min={0}
				max={80}
				value={item.margin}
				onChange={val => props.setItem({...props.item, margin: val})}
			/>

			<Property4SideRange
				id='Frame-padding'
				label={Trans('padding')}
				min={0}
				max={80}
				value={item.padding}
				onChange={val => props.setItem({...props.item, padding: val})}
			/>

			<div className='section-name'>
				{Trans('border')}
			</div>

			<div className='form-check mb-3'>
				<input
					type='checkbox'
					id='Frame-border-single'
					className='form-check-input'
					checked={Array.isArray(item.border)}
					onChange={e => {
						const obj: FrameData = {...item}
						if (e.currentTarget.checked) {
							const brd: Border = !Array.isArray(item.border) ? item.border : {width:1, color: '#cccccc', style:'solid'}
							obj.border = [{...brd}, {...brd}, {...brd}, {...brd}]
						}
						else {
							obj.border = {width:1, color: '#cccccc', style:'solid'}
						}
						props.setItem(obj)
					}}
				/>
				<label className="form-check-label" htmlFor='Frame-border-single'>
					{Trans('border different sides')}
				</label>
			</div>
			
			{Array.isArray(item.border) && ['top','right','bottom','left'].map((side2, idx) => {
				const side = side2 as 'top'|'right'|'bottom'|'left'
				const val2 = Array.isArray(item.border) ? item.border : [item.border, item.border, item.border, item.border]
				return <React.Fragment key={side}>
					<div className='section-name'>
						{Trans(`border-${side}`)}
					</div>
					<PropertyBorder
						id={`Frame-border-${side}`}
						value={val2[idx]}
						onChange={val => {
							const arr = [...val2]
							arr[idx] = val
							props.setItem({...item, border: arr})
						}}
					/>
				</React.Fragment>
			})}
			{!Array.isArray(item.border) && <PropertyBorder
				id={`Frame-border-full`}
				value={item.border}
				onChange={val => props.setItem({...item, border: val})}
			/>}

			<div className='section-name'>
				{Trans('other')}
			</div>
			<div className='hform'>
				<label>
					{Trans('font')}
				</label>
				<PropertyFont
					value={item.font}
					onChange={val => props.setItem({...props.item, font: val})}
					googleFontApiKey={props.api.googleFontApiKey}
				/>
			</div>
			<div className='form-check'>
				<input
					type='checkbox'
					id='Frame-page-break-avoid'
					className='form-check-input'
					checked={!!item.pageBreakAvoid}
					onChange={e => {
						const obj: FrameData = {...item}
						obj.pageBreakAvoid = e.currentTarget.checked
						props.setItem(obj)
					}}
				/>
				<label className='form-check-label' htmlFor='Frame-page-break-avoid'>
					{Trans('page-break-avoid')}
				</label>
			</div>
			
		</>
	},
}
