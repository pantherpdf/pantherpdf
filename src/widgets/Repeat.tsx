/**
 * Repeat.tsx
 */

import React, { CSSProperties } from 'react'
import { TData, TDataCompiled, tuple } from '../types'
import type { Widget } from '../editor/types'
import { faEllipsisV } from '@fortawesome/free-solid-svg-icons'
import BoxName from './BoxName'
import InputApplyOnEnter from './InputApplyOnEnter'
import Trans, { TransName, trKeys } from '../translation'


const RepeatDirections = tuple('rows', 'columns', 'grid');
type RepeatDirection = (typeof RepeatDirections)[number];
const RepeatDirectionTrans: {[key in RepeatDirection]: trKeys} = {
	rows: 'repeat - direction rows',
	columns: 'repeat - direction columns',
	grid: 'repeat - direction grid',
}

export interface RepeatData extends TData {
	type: 'Repeat',
	source: string,
	varName: string,
	direction: RepeatDirection,
}

export interface RepeatCompiled extends TDataCompiled {
	type: 'Repeat',
	children: TDataCompiled[][],
	direction: RepeatDirection,
	addChildElement: boolean,
}


export const Repeat: Widget = {
	name: {en: 'Repeat', sl: 'Ponavljaj'},
	icon: {fontawesome: faEllipsisV},

	newItem: async (): Promise<RepeatData> => {
		return {
			type: 'Repeat',
			children: [],
			source: '',
			varName: 'item',
			direction: 'rows',
		}
	},

	compile: async (dt: RepeatData, helper): Promise<RepeatCompiled> => {
		const value = await helper.evalFormula(dt.source)
		if (!Array.isArray(value)) {
			throw new Error(`Repeat: expected source to be array but got ${typeof value}`)
		}
		const children: TDataCompiled[][] = []
		for (let i = 0; i < value.length; ++i) {
			helper.formulaHelper.push(dt.varName, value[i])
			helper.formulaHelper.push(dt.varName+'_i', i)
			const ch2 = await helper.compileChildren(dt.children, helper)
			children.push(ch2)
			helper.formulaHelper.pop()
			helper.formulaHelper.pop()
		}
		let addChildElement = true
		if (dt.direction === 'grid') {
			// remove <div> when only one child and this child is frame
			if (dt.children.length === 1 && dt.children[0].type === 'Frame') {
				// dont display as flex because flex and page-break-inside: avoid dont work together
				// dont add additional div, so that frame can control width/height
				// add css to frame
				// inline-block to display them in line
				// vertical-align: top, to prevent vertical space between frames
				addChildElement = false
				if (helper.reportCompiled.globalCss.indexOf('.grid-with-frame') === -1) {
					helper.reportCompiled.globalCss += `
						.grid-with-frame > div {
							display: inline-block;
							vertical-align: top;
						}
					`
				}
			}
		}
		return {
			type: dt.type,
			children,
			direction: dt.direction,
			addChildElement,
		}
	},

	Render: function(props) {
		const item = props.item as RepeatData
		return <BoxName {...props} name={`${TransName(Repeat.name)} - ${item.varName}`}>
			{props.item.children && props.renderWidgets(props.item.children, props.wid)}
		</BoxName>
	},

	RenderFinal: function(props) {
		const item = props.item as RepeatCompiled
		if (item.direction === 'rows') {
			// maybe add to cssItem:
			// pageBreakInside: 'avoid',  // must be present otherwise item in quotes gets breaked up
			return item.children.map(item => props.renderChildren(item, props)).join('') + '\n'
		}
		const cssParent: CSSProperties = {}
		const cssItem: CSSProperties = {}
		if (item.direction === 'columns') {
			cssParent.display = 'flex'
			const w = item.children.length > 0 ? 1 / item.children.length : 1
			cssItem.flex = `0 0 ${(w * 100).toLocaleString('en-US', {maximumFractionDigits:4})}%`
		}
		if (item.direction === 'grid') {
			if (item.addChildElement) {
				cssParent.display = 'flex'
				cssParent.flexWrap = 'wrap'
			}
			else {
				cssParent.display = 'block'
			}
		}

		if (item.addChildElement) {
			return `<div style="${props.styleToStringAttribute(cssParent)}">
				${item.children.map((item2, idx) => {
					return `<div
						style="${props.styleToStringAttribute(cssItem)}"
					>
						${props.renderChildren(item2, props)}
					</div>`
				}).join('')}
			</div>\n`
		}
		else {
			return `<div style="${props.styleToStringAttribute(cssParent)}" class="grid-with-frame">
				${item.children.map(item2 => props.renderChildren(item2, props)).join('')}
			</div>\n`
		}
	},

	RenderProperties: function(props) {
		const item = props.item as RepeatData
		return <>
			<div className='hform'>
				<label htmlFor='Repeat-source'>
					{Trans('source data')}
				</label>
				<div className='input-group'>
					<span className="input-group-text fst-italic">Æ’</span>
					<InputApplyOnEnter
						id='Repeat-source'
						value={item.source}
						onChange={val=>props.setItem({...item, source: val})}
					/>
				</div>
			</div>

			<div className='hform'>
				<label htmlFor='Repeat-varName'>
					{Trans('varName')}
				</label>
				<InputApplyOnEnter
					id='Repeat-varName'
					value={item.varName}
					onChange={val=>props.setItem({...item, varName: val})}
				/>
			</div>
			<small className='text-muted d-block'>
				{Trans('repeat - current item is this var')}
			</small>
			<small className='text-muted d-block mb-3'>
				{Trans('repeat - index name', [`${item.varName}_i`])}
			</small>

			<div className='hform'>
				<label htmlFor='Repeat-direction'>
					{Trans('repeat - direction')}
				</label>
				<select
					className='form-select'
					value={item.direction}
					onChange={e => props.setItem({...item, direction: e.currentTarget.value})}
					id='Repeat-direction'
				>
					{RepeatDirections.map(m => <option
						key={m}
						value={m}
					>
						{Trans(RepeatDirectionTrans[m])}
					</option>)}
				</select>
			</div>
		</>
	},
}
