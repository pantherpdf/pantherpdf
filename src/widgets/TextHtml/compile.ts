import { listOfAdjusts } from '../formulaAdjust';
import { CompileHelper, WidgetItem } from '../types';
import type { TextHtmlCompiled, TextHtmlData } from './TextHtml';

export default async function compile(
  item: WidgetItem,
  helper: CompileHelper,
): Promise<TextHtmlCompiled> {
  const dt = item as TextHtmlData;
  // combine parts
  let html = '';
  for (const part of dt.value) {
    const type = part.type;
    switch (type) {
      case 'html': {
        html += part.value;
        break;
      }
      case 'formula': {
        let value = await helper.evalFormula(part);
        if (part.adjust && part.adjust.length > 0) {
          const adjustObj = listOfAdjusts.find(x => x.id === part.adjust);
          if (!adjustObj) {
            throw new Error(`Unknown adjust ${part.adjust}`);
          }
          value = adjustObj.func(value, []);
        }
        html += String(value);
        break;
      }
      default: {
        const exhaustiveCheck: never = type;
        throw new Error(`Unknown data type: ${exhaustiveCheck}`);
      }
    }
  }
  return {
    type: dt.type,
    value: html,
    font: dt.font,
  };
}
