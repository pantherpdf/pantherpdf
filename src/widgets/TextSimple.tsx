/**
 * TextSimple.tsx
 */

import React from 'react';
import { Item, ItemCompiled } from '../types';
import type { Widget } from '../editor/types';
import { faAlignLeft } from '@fortawesome/free-solid-svg-icons';
import BoxName from './BoxName';
import InputApplyOnEnter from './InputApplyOnEnter';

export interface TextSimpleData extends Item {
  type: 'TextSimple';
  formula: string;
}

export interface TextSimpleCompiled extends ItemCompiled {
  type: 'TextSimple';
  data: string;
}

export const TextSimple: Widget = {
  id: 'TextSimple',
  name: { en: 'Text Simple', sl: 'Besedilo Simple' },
  icon: faAlignLeft,

  newItem: async (): Promise<TextSimpleData> => {
    return {
      type: 'TextSimple',
      children: [],
      formula: '',
    };
  },

  compile: async (dt: TextSimpleData, helpers): Promise<TextSimpleCompiled> => {
    const str2 = await helpers.evalFormula(dt.formula);
    const str =
      str2 !== undefined && str2 !== null && str2 !== false ? String(str2) : '';
    return {
      type: dt.type,
      data: str,
    };
  },

  RenderEditor: function (props) {
    const item = props.item as TextSimpleData;
    return (
      <BoxName {...props} name={TextSimple.name}>
        <div>{item.formula}</div>
      </BoxName>
    );
  },

  RenderPreview: function (props) {
    const item = props.item as TextSimpleCompiled;
    return `<div>${props.escapeHtml(item.data)}</div>\n`;
  },

  RenderProperties: function (props) {
    const item = props.item as TextSimpleData;
    return (
      <>
        <div className="input-group mb-3">
          <span className="input-group-text fst-italic">Æ’</span>
          <InputApplyOnEnter
            id="TextSimple-formula"
            value={item.formula}
            onChange={val => props.setItem({ ...item, formula: val })}
          />
        </div>
      </>
    );
  },

  canAdd: false,
};
