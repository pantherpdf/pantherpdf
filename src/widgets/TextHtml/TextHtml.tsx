/**
 * @project PantherPDF Report Editor
 * @copyright Ignac Banic 2021-2024
 * @license MIT
 */

// manual tests:
// - widget not selected, try to drag it. It should not be draggable, however it should select text
// - type some text, then wait few sec to periodicaly save, the cursor should stay on same place
// - insert tag, delete nbsp space after tag and press enter

import React from 'react';
import type { WidgetItem, WidgetCompiled, Widget } from '../types';
import { PropertyFontGenCss, TFont } from '../PropertyFont';
import { faAlignLeft } from '@fortawesome/free-solid-svg-icons';
import trans from '../../translation';
import PropertiesEditor from './PropertiesEditor';
import { TextHtmlDataValue } from './internalRepresentation';
import { escapeHtml } from './helpers';
import { name } from './options';
import compile from './compile';
import EditorWrapper from './EditorWrapper';

export interface TextHtmlData extends WidgetItem {
  type: 'TextHtml';
  value: TextHtmlDataValue[];
  font: TFont;
}

export interface TextHtmlCompiled extends WidgetCompiled {
  type: 'TextHtml';
  value: string;
  font: TFont;
}

export const TextHtml: Widget = {
  id: 'TextHtml',
  name,
  icon: faAlignLeft,

  newItem: async (props): Promise<TextHtmlData> => {
    const valueTxt = `<div>${escapeHtml(
      trans('TextHtml initial value'),
    )}</div>`;
    const value: TextHtmlDataValue[] =
      props.report.widgets.length > 0
        ? [{ type: 'html', value: valueTxt }]
        : [];
    return {
      type: 'TextHtml',
      children: [],
      value,
      font: {},
    };
  },

  compile,

  Editor: EditorWrapper,

  Preview: function (props) {
    const item = props.item as TextHtmlCompiled;
    const css = PropertyFontGenCss(item.font);
    return <div style={css} dangerouslySetInnerHTML={{ __html: item.value }} />;
  },

  Properties: PropertiesEditor,

  canDrag: false,
};
