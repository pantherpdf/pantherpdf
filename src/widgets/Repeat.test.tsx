/**
 * @jest-environment jsdom
 */

import { TextSimpleData } from './TextSimple'
import { RepeatData } from './Repeat'
import compile, { compileComponent } from '../editor/compile'
import { ForceChildren } from '../editor/types'
import { sampleReport } from '../editor/sampleReport'
import { TReport } from '../types'
import { TextHtmlData } from './TextHtml'
import { makeHtmlContent } from '../editor/makeHtml'
import renderer from 'react-test-renderer'
import { FrameData } from './Frame'


test('text', async () => {
	const dt: ForceChildren<RepeatData|TextSimpleData> = {type:'Repeat', varName:'rp', source:'data.arr + ["a","b"]', direction:'rows', children:[
		{type:'TextSimple', formula:'rp', children:[]}
	]}
	const data = { arr:['1','2'] }
	const p = await compileComponent(dt, data) as any
	expect(p).toBeTruthy()
	expect(p.type).toBe('Repeat')
	expect(p.children[0][0].data).toBe('1')
	expect(p.children[1][0].data).toBe('2')
	expect(p.children[2][0].data).toBe('a')
	expect(p.children[3][0].data).toBe('b')
})


test('complete rows', async () => {
	const txt: TextHtmlData = { type: 'TextHtml', value: '<data>item</data>', font:{}, children:[] }
	const rpt: RepeatData = { type: 'Repeat', children: [txt], source: '[1,2,3,4,5]', varName: 'item', direction: 'rows' }
	const report: TReport = {...sampleReport, children: [rpt] }
	const report2 = await compile(report, {})
	
	const children = report2.children as any
	expect(children[0].type).toBe('Repeat')
	expect(children[0].children.length).toBe(5)
	expect(Array.isArray(children[0].children[0])).toBeTruthy()
	expect(children[0].children[0][0].type).toBe('TextHtml')
	expect(children[0].children[0][0].value).toBe('1')

	const html = makeHtmlContent(report2)
	const component = renderer.create(<>{html}</>)
	const tree = component.toJSON()
  	expect(tree).toMatchSnapshot();
})


test('complete grid', async () => {
	const txt: TextHtmlData = { type: 'TextHtml', value: '<data>item</data>', font:{}, children:[] }
	const rpt: RepeatData = { type: 'Repeat', children: [txt], source: '[1,2,3,4,5]', varName: 'item', direction: 'grid' }
	const report: TReport = {...sampleReport, children: [rpt] }
	const report2 = await compile(report, {})

	const children = report2.children as any
	expect(children[0].type).toBe('Repeat')
	expect(children[0].children.length).toBe(5)
	expect(Array.isArray(children[0].children[0])).toBeTruthy()
	expect(children[0].children[0][0].type).toBe('TextHtml')
	expect(children[0].children[0][0].value).toBe('1')

	const html = makeHtmlContent(report2)
	const component = renderer.create(<>{html}</>)
	const tree = component.toJSON()
  	expect(tree).toMatchSnapshot();
})


test('grid - products', async () => {
	const frame: FrameData = {
		type: 'Frame',
		children: [],
		margin: [0, 0, 0, 0],
		padding: [0, 0, 0, 0],
		border: {
			width: 0,
			color: '',
			style: 'solid',
		},
		width: '5cm',
		height: '5cm',
		font: {},
	}
	const rpt: RepeatData = { type: 'Repeat', children: [frame], source: '[1,2,3]', varName: 'item', direction: 'grid' }
	const report: TReport = {...sampleReport, children: [rpt] }
	const report2 = await compile(report, {})

	const html = makeHtmlContent(report2)
	const component = renderer.create(<>{html}</>)
	const tree = component.toJSON()
  	expect(tree).toMatchSnapshot();
	expect(report2.globalCss.replace(/\s/g,'')).toContain('.grid-with-frame>div{display:inline-block;vertical-align:top;}')
})
