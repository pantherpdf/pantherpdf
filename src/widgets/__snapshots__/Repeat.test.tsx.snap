// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`complete grid 1`] = `
<div
  style={
    Object {
      "display": "flex",
      "flexWrap": "wrap",
    }
  }
>
  <div
    style={Object {}}
  >
    <div
      dangerouslySetInnerHTML={
        Object {
          "__html": "1",
        }
      }
      style={Object {}}
    />
  </div>
  <div
    style={Object {}}
  >
    <div
      dangerouslySetInnerHTML={
        Object {
          "__html": "2",
        }
      }
      style={Object {}}
    />
  </div>
  <div
    style={Object {}}
  >
    <div
      dangerouslySetInnerHTML={
        Object {
          "__html": "3",
        }
      }
      style={Object {}}
    />
  </div>
  <div
    style={Object {}}
  >
    <div
      dangerouslySetInnerHTML={
        Object {
          "__html": "4",
        }
      }
      style={Object {}}
    />
  </div>
  <div
    style={Object {}}
  >
    <div
      dangerouslySetInnerHTML={
        Object {
          "__html": "5",
        }
      }
      style={Object {}}
    />
  </div>
</div>
`;

exports[`complete rows 1`] = `
Array [
  <div
    dangerouslySetInnerHTML={
      Object {
        "__html": "1",
      }
    }
    style={Object {}}
  />,
  <div
    dangerouslySetInnerHTML={
      Object {
        "__html": "2",
      }
    }
    style={Object {}}
  />,
  <div
    dangerouslySetInnerHTML={
      Object {
        "__html": "3",
      }
    }
    style={Object {}}
  />,
  <div
    dangerouslySetInnerHTML={
      Object {
        "__html": "4",
      }
    }
    style={Object {}}
  />,
  <div
    dangerouslySetInnerHTML={
      Object {
        "__html": "5",
      }
    }
    style={Object {}}
  />,
]
`;

exports[`grid - products 1`] = `
<div
  className="grid-with-frame"
  style={
    Object {
      "display": "block",
    }
  }
>
  <div
    style={
      Object {
        "border": "0px solid ",
        "boxSizing": "border-box",
        "flex": "0 0 5cm",
        "height": "5cm",
        "margin": "0px 0px 0px 0px",
        "overflowX": "hidden",
        "overflowY": "hidden",
        "padding": "0px 0px 0px 0px",
        "width": "5cm",
      }
    }
  />
  <div
    style={
      Object {
        "border": "0px solid ",
        "boxSizing": "border-box",
        "flex": "0 0 5cm",
        "height": "5cm",
        "margin": "0px 0px 0px 0px",
        "overflowX": "hidden",
        "overflowY": "hidden",
        "padding": "0px 0px 0px 0px",
        "width": "5cm",
      }
    }
  />
  <div
    style={
      Object {
        "border": "0px solid ",
        "boxSizing": "border-box",
        "flex": "0 0 5cm",
        "height": "5cm",
        "margin": "0px 0px 0px 0px",
        "overflowX": "hidden",
        "overflowY": "hidden",
        "padding": "0px 0px 0px 0px",
        "width": "5cm",
      }
    }
  />
</div>
`;
